src/provider/view/utils/Grouping.ts

```diff
@@ -0,0 +1,108 @@
+ * these groups.
+ */
+export interface Group {
+    /**
+     * Unique key for this group. Mainly used for storing a date string so as time passes we do not need to recalculate the date to group the notifications by.
+     */
+    key: string;
+    /**
+     * The grouping type used to create this group.
+     */
+    type: GroupingType;
+    /**
+     * User-visibile title that will be displayed above these notifications
+     */
+    title: string;
+
+    /**
+     * The notifications to display within this group.
+     *
+     * This array should always contain at least one notification. If there are no
+     * notifications that fit within this category then the group should be deleted.
+     */
+    notifications: StoredNotification[];
+}
+
+/**
+ * When creating a new notification group, this function determines the
+ * user-visible title for the group
+ * @param notification Notificiation
+ * @param groupingType Grouping method
+ */
+export function getGroupTitle(notification: StoredNotification, groupingType: GroupingType): string {
+    if (groupingType === GroupingType.APPLICATION) {
+        return notification.source.name || notification.source.uuid;
+    } else if (groupingType === GroupingType.DATE) {
+        const {date} = notification.notification;
+        return moment(date).calendar(undefined, {
+            sameDay: '[Today]',
+            nextDay: '[Tomorrow]',
+            nextWeek: 'dddd',
+            lastDay: '[Yesterday]',
+            lastWeek: '[Last] dddd',
+            sameElse: 'DD/MM/YYYY'
+        });
+    }
+    throw new Error(`invalid groupMethod : ${groupingType}`);
+}
+
+/**
+ * Group notifications together based on the given grouping method e.g. application/date.
+ * @param notifications List of notifications to sort into groups.
+ * @param groupMethod Grouping type to use for sorting.
+ */
+export function groupNotifications(notifications: StoredNotification[], groupMethod: GroupingType): Map<string, Group> {
+    // TODO: This sorting should be removed and changed to use insert-sort inside the reduce below
+    // Pre-sort notifications by date (groups will then also be sorted by date)
+    notifications.sort((a: StoredNotification, b: StoredNotification) =>
+        b.notification.date.valueOf() - a.notification.date.valueOf());
+    // Reduce the notifications array into a Map of notifications grouped by their title
+    const groupMap = notifications.reduce((groups: Map<string, Group>, currentNotification: StoredNotification) => {
+        const groupTitle = getGroupTitle(currentNotification, groupMethod);
+        // If group title already exists just add it to the group
+        if (groups.has(groupTitle)) {
+            const group = groups.get(groupTitle)!;
+            groups.set(groupTitle, {
+                ...group,
+                notifications: [
+                    ...group.notifications,
+                    currentNotification
+                ]
```

Comment:
Do we need to get, push and set here? Would've thought a `group.notifications.push(currentNotification)` would suffice?

There's no kind of react change-detection issue here, as the whole Group object would only have been created within a previous iteration of this reduce operation?

Reply:
Switched to push. There was a change detection issue with keys as the array was the same, but added `notification.date` to the key so it would be unique and rerender.

Reply:
Also just realised we don't need to even use sort, we can just `reduceRight` since notifications are in chronological order.